metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py

```diff
@@ -0,0 +1,341 @@
+                    "%s users would not get ingested as admin permission is not enabled on "
+                    "configured Azure AD Application",
+                    entity_name,
+                )
+                return users
+            # if Other error then re-raise
+            raise e
+
+        return users
+
+    def get_dashboard_users(self, dashboard: Dashboard) -> List[User]:
+        return self._get_entity_users(
+            dashboard.workspace_id, "dashboards", dashboard.id
+        )
+
+    def get_report_users(self, workspace_id: str, report_id: str) -> List[User]:
+        return self._get_entity_users(workspace_id, "reports", report_id)
+
+    def get_reports(self, workspace: Workspace) -> List[Report]:
+        """
+        Fetch the report from PowerBi for the given Workspace
+        """
+        if workspace is None:
+            logger.info("workspace is None")
+            return []
+
+        reports: List[Report] = self._get_resolver().get_reports(workspace)
+
+        def fill_ownership() -> None:
+            if self.__config.extract_ownership is False:
+                logger.info(
+                    "Skipping user retrieval for report as extract_ownership is set to false"
+                )
+                return
+
+            for report in reports:
+                report.users = self.get_report_users(
+                    workspace_id=workspace.id, report_id=report.id
+                )
+
+        def fill_tags() -> None:
+            if self.__config.extract_endorsements_to_tags is False:
+                logger.info(
+                    "Skipping endorsements tags retrieval for report as extract_endorsements_to_tags is set to false"
+                )
+                return
+
+            for report in reports:
+                report.tags = workspace.report_endorsements.get(report.id, [])
+
+        fill_ownership()
+        fill_tags()
+
+        return reports
+
+    def get_workspaces(self) -> List[Workspace]:
+        groups: List[dict] = self._get_resolver().get_groups()
+        workspaces = [
+            Workspace(
+                id=workspace["id"],
+                name=workspace["name"],
+                datasets={},
+                dashboards=[],
+                reports=[],
+                report_endorsements={},
+                dashboard_endorsements={},
+                scan_result={},
+            )
+            for workspace in groups
+        ]
+        return workspaces
+
+    def _get_scan_result(self, workspace: Workspace) -> Any:
+        scan_id: Optional[str] = None
+        try:
+            scan_id = self.__admin_api_resolver.create_scan_job(
+                workspace_id=workspace.id
+            )
+        except requests.exceptions.HTTPError as e:
+            if data_resolver.is_permission_error(e):
+                logger.warning(
+                    "Lineage would not get ingested as admin permission is not enabled on configured Azure AD "
```

Comment:
Lineage between Dashboard and Dataset? .. lets add that text

Reply:
Minor thing: 

"Lineage can not be ingestion because this user does not have access to the Tableau Admin API."

Reply:
Done